File: frontend/src/utils/TaskManager.js
/**
 * Task Management System for Inference and Training Tasks
 * Handles task CRUD operations, persistence, and queue management
 */

import { v4 as uuidv4 } from 'uuid';

export const TASK_STATUS = {
  UNSTARTED: 'unstarted',
  QUEUED: 'queued',
  RUNNING: 'running',
  COMPLETED: 'completed',
  FAILED: 'failed',
  CANCELLED: 'cancelled'
};

export const TASK_TYPE = {
  INFERENCE: 'inference',
  TRAINING: 'training'
};

class TaskManager {
  constructor() {
    this.tasks = new Map();
    this.queue = [];
    this.currentTask = null;
    this.listeners = new Set();
    this.loadTasks();
  }

  // Task CRUD Operations
  createTask(config, name = null, taskType = null) {
    // Determine task type from config or parameter
    const type = taskType || this.determineTaskType(config);

    const task = {
      id: uuidv4(),
      name: name || this.generateTaskName(config, type),
      type: type,
      status: TASK_STATUS.UNSTARTED,
      config: { ...config },
      created: Date.now(),
      started: null,
      completed: null,
      progress: '',
      result: null,
      processId: null,
      systemPid: null
    };

    this.tasks.set(task.id, task);
    this.saveTasks();
    this.notifyListeners('taskCreated', task);
    return task;
  }

  determineTaskType(config) {
    // Check if config has training-specific properties
    if (config.class_list !== undefined ||
      config.fully_annotated_files !== undefined ||
      config.single_class_annotations !== undefined) {
      return TASK_TYPE.TRAINING;
    }
    return TASK_TYPE.INFERENCE;
  }

  getTask(taskId) {
    return this.tasks.get(taskId);
  }

  getAllTasks() {
    return Array.from(this.tasks.values()).sort((a, b) => b.created - a.created);
  }

  updateTask(taskId, updates) {
    const task = this.tasks.get(taskId);
    if (!task) return null;

    Object.assign(task, updates);
    this.tasks.set(taskId, task);
    this.saveTasks();
    this.notifyListeners('taskUpdated', task);
    return task;
  }

  deleteTask(taskId) {
    const task = this.tasks.get(taskId);
    if (!task) return false;

    // Can't delete running tasks
    if (task.status === TASK_STATUS.RUNNING) {
      return false;
    }

    // Remove from queue if present
    this.queue = this.queue.filter(id => id !== taskId);

    this.tasks.delete(taskId);
    this.saveTasks();
    this.notifyListeners('taskDeleted', { id: taskId });
    return true;
  }

  // Queue Management
  queueTask(taskId) {
    const task = this.getTask(taskId);
    if (!task || task.status === TASK_STATUS.RUNNING) return false;

    // Update task status to queued
    this.updateTask(taskId, {
      status: TASK_STATUS.QUEUED,
      progress: 'Queued for execution'
    });

    // Add to queue if not already present
    if (!this.queue.includes(taskId)) {
      this.queue.push(taskId);
    }

    // Start immediately if no task is running
    if (!this.currentTask) {
      this.processQueue();
    }

    return true;
  }

  async processQueue() {
    if (this.currentTask || this.queue.length === 0) return;

    const taskId = this.queue.shift();
    const task = this.getTask(taskId);

    if (!task || task.status === TASK_STATUS.CANCELLED) {
      // Task was deleted or cancelled, process next
      this.processQueue();
      return;
    }

    this.currentTask = taskId;
    await this.executeTask(taskId);
  }

  async executeTask(taskId) {
    const task = this.getTask(taskId);
    if (!task) return;

    try {
      // Update task to running status with appropriate start message
      const startMessage = task.type === TASK_TYPE.TRAINING ? 'Starting training...' : 'Starting inference...';
      this.updateTask(taskId, {
        status: TASK_STATUS.RUNNING,
        started: Date.now(),
        progress: startMessage
      });

      // Execute based on task type
      const result = task.type === TASK_TYPE.TRAINING
        ? await this.runTraining(task)
        : await this.runInference(task);

      // Check if task was cancelled
      if (result && result.cancelled) {
        // Task was cancelled, update accordingly
        this.updateTask(taskId, {
          status: TASK_STATUS.CANCELLED,
          completed: Date.now(),
          progress: 'Cancelled by user',
          result: result
        });
      } else {
        // Update task with results
        this.updateTask(taskId, {
          status: TASK_STATUS.COMPLETED,
          completed: Date.now(),
          progress: 'Completed successfully',
          result: result
        });
      }

    } catch (error) {
      // Update task with error
      this.updateTask(taskId, {
        status: TASK_STATUS.FAILED,
        completed: Date.now(),
        progress: 'Failed',
        result: { error: error.message }
      });
    } finally {
      // Clear current task and process next in queue
      this.currentTask = null;
      setTimeout(() => this.processQueue(), 100);
    }
  }

  async runInference(task) {
    const config = task.config;

    // Generate unique process ID for this task
    const processId = `task_${task.id}_${Date.now()}`;

    // Update task with process ID
    this.updateTask(task.id, { processId });

    try {
      // Update progress
      this.updateTask(task.id, { progress: 'Preparing configuration...' });

      // Create job folder and output paths with unique name
      const jobFolderName = await this.generateUniqueJobFolderName(config.output_dir, task.name);
      const jobFolder = config.output_dir ? `${config.output_dir}/${jobFolderName}` : '';

      // Determine output file based on sparse threshold setting
      const isSparseOutput = config.sparse_save_threshold !== null && config.sparse_save_threshold !== undefined;
      const outputFile = jobFolder ? `${jobFolder}/${isSparseOutput ? 'sparse_predictions.pkl' : 'predictions.csv'}` : '';

      const configJsonPath = jobFolder ? `${jobFolder}/${task.name}_${task.id}.json` : '';
      const logFilePath = jobFolder ? `${jobFolder}/inference_log.txt` : '';

      // Create temporary config file
      const tempConfigPath = `/tmp/inference_config_${processId}.json`;
      const configData = {
        model_source: config.model_source || 'bmz',
        model: config.model,
        files: config.files || [],
        file_globbing_patterns: config.file_globbing_patterns || [],
        file_list: config.file_list || '',
        output_dir: config.output_dir,
        output_file: outputFile,
        sparse_save_threshold: config.sparse_save_threshold || null,
        job_folder: jobFolder,
        config_output_path: configJsonPath,
        log_file_path: logFilePath,
        split_by_subfolder: config.split_by_subfolder || false,
        subset_size: config.testing_mode_enabled ? config.subset_size : null,
        inference_settings: {
          clip_overlap: config.overlap || 0.0,
          batch_size: config.batch_size || 1,
          num_workers: config.worker_count || 1
        }
      };

      // Save temporary config file using HTTP API
      const saveResponse = await fetch('http://localhost:8000/config/save', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          config_data: configData,
          output_path: tempConfigPath
        })
      });

      const saveResult = await saveResponse.json();
      if (saveResult.status !== 'success') {
        throw new Error(`Failed to save configuration: ${saveResult.error}`);
      }

      // Update progress
      this.updateTask(task.id, { progress: 'Setting up ML environment...' });

      // Get environment paths - use custom environment if specified
      let envPath, archivePath;
      if (config.use_custom_python_env && config.custom_python_env_path) {
        envPath = config.custom_python_env_path;
        archivePath = ''; // No archive needed for custom environments
      } else {
        // Use default environment
        const envPathResult = await window.electronAPI.getEnvironmentPath('dipper_pytorch_env');
        const archivePathResult = await window.electronAPI.getArchivePath('dipper_pytorch_env.tar.gz');

        if (!envPathResult.success || !archivePathResult.success) {
          throw new Error('Failed to get environment paths');
        }

        envPath = envPathResult.path;
        archivePath = archivePathResult.path;
      }

      // Update progress
      this.updateTask(task.id, { progress: 'Running inference...' });

      // Start inference via HTTP API (now returns immediately with job ID)
      const inferenceResponse = await fetch('http://localhost:8000/inference/run', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          config_path: tempConfigPath,
          env_path: envPath,
          archive_path: archivePath,
          job_id: processId
        })
      });

      const startResult = await inferenceResponse.json();

      if (startResult.status === 'started') {
        const jobId = startResult.job_id;
        this.updateTask(task.id, { progress: 'Inference running...', processId: jobId });

        // Poll for completion
        const finalResult = await this.pollInferenceStatus(jobId, task.id);

        if (finalResult.status === 'completed') {
          return {
            success: true,
            output_path: configData.output_file,
            job_folder: configData.job_folder,
            config_path: configData.config_output_path,
            total_files: config.files.length || config.file_globbing_patterns.length || (config.file_list ? 1 : 0),
            message: 'Inference completed successfully',
            job_id: jobId
          };
        } else if (finalResult.status === 'cancelled') {
          return {
            success: false,
            cancelled: true,
            message: 'Inference was cancelled by user',
            job_id: jobId
          };
        } else {
          // Show detailed error information
          let errorMessage = finalResult.error || 'Inference failed';

          if (finalResult.stderr) {
            errorMessage += '\n\nError output:\n' + finalResult.stderr;
          }

          if (finalResult.stdout) {
            errorMessage += '\n\nStandard output:\n' + finalResult.stdout;
          }

          throw new Error(errorMessage);
        }
      } else {
        throw new Error(startResult.error || 'Failed to start inference');
      }
    } catch (error) {
      throw new Error(`Inference failed: ${error.message}`);
    }
  }

  async pollInferenceStatus(jobId, taskId, pollInterval = 2000) {
    /**
     * Poll inference status until completion or failure
     * @param {string} jobId - Backend job ID 
     * @param {string} taskId - Frontend task ID
     * @param {number} pollInterval - Polling interval in milliseconds
     * @returns {Promise} Final result when inference completes
     */
    return new Promise((resolve, reject) => {
      const poll = async () => {
        try {
          const response = await fetch(`http://localhost:8000/inference/status/${jobId}`);

          if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${response.statusText}`);
          }

          const result = await response.json();

          // Update task progress based on status
          if (result.status === 'running') {
            // Store system PID if available
            const updates = { progress: 'Inference running...' };
            if (result.system_pid) {
              updates.systemPid = result.system_pid;
            }
            this.updateTask(taskId, updates);
            // Continue polling
            setTimeout(poll, pollInterval);
          } else if (result.status === 'completed') {
            this.updateTask(taskId, { progress: 'Inference completed' });
            resolve(result);
          } else if (result.status === 'failed') {
            this.updateTask(taskId, { progress: 'Inference failed' });
            resolve(result); // Resolve with failed status, let caller handle error
          } else if (result.status === 'cancelled') {
            this.updateTask(taskId, { progress: 'Cancelled by user' });
            resolve(result); // Resolve with cancelled status - stop polling
          } else {
            // Unknown status, treat as error
            reject(new Error(`Unknown inference status: ${result.status}`));
          }
        } catch (error) {
          reject(new Error(`Failed to check inference status: ${error.message}`));
        }
      };

      // Start polling
      poll();
    });
  }

  async runTraining(task) {
    const config = task.config;

    // Generate unique process ID for this task
    const processId = `training_${task.id}_${Date.now()}`;

    // Update task with process ID
    this.updateTask(task.id, { processId });

    try {
      // Update progress
      this.updateTask(task.id, { progress: 'Preparing training configuration...' });

      // Create job folder and output paths with unique name
      const jobFolderName = await this.generateUniqueJobFolderName(config.save_location, task.name);
      const jobFolder = config.save_location ? `${config.save_location}/${jobFolderName}` : '';
      const modelSavePath = jobFolder ? `${jobFolder}/trained_model.pth` : '';
      const configJsonPath = jobFolder ? `${jobFolder}/${task.name}_${task.id}.json` : '';
      const logFilePath = jobFolder ? `${jobFolder}/train_log.txt` : '';

      // Create temporary config file
      const tempConfigPath = `/tmp/training_config_${processId}.json`;
      const configData = {
        model: config.model || 'BirdNET',
        class_list: config.class_list || [],
        fully_annotated_files: config.fully_annotated_files || [],
        single_class_annotations: config.single_class_annotations || [],
        background_samples_file: config.background_samples_file || '',
        root_audio_folder: config.root_audio_folder || '',
        evaluation_file: config.evaluation_file || '',
        save_location: config.save_location,
        model_save_path: modelSavePath,
        job_folder: jobFolder,
        config_output_path: configJsonPath,
        log_file_path: logFilePath,
        subset_size: config.testing_mode_enabled ? config.subset_size : null,
        training_settings: {
          batch_size: config.batch_size || 32,
          num_workers: config.num_workers || 4,
          freeze_feature_extractor: config.freeze_feature_extractor !== false,
          classifier_hidden_layer_sizes: config.training_settings?.classifier_hidden_layer_sizes || null
        }
      };

      // Save temporary config file using HTTP API
      const saveResponse = await fetch('http://localhost:8000/config/save', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          config_data: configData,
          output_path: tempConfigPath
        })
      });

      const saveResult = await saveResponse.json();
      if (saveResult.status !== 'success') {
        throw new Error(`Failed to save configuration: ${saveResult.error}`);
      }

      // Update progress
      this.updateTask(task.id, { progress: 'Setting up ML environment...' });

      // Get environment paths - use custom environment if specified
      let envPath, archivePath;
      if (config.use_custom_python_env && config.custom_python_env_path) {
        envPath = config.custom_python_env_path;
        archivePath = ''; // No archive needed for custom environments
      } else {
        // Use default environment
        const envPathResult = await window.electronAPI.getEnvironmentPath('dipper_pytorch_env');
        const archivePathResult = await window.electronAPI.getArchivePath('dipper_pytorch_env.tar.gz');

        if (!envPathResult.success || !archivePathResult.success) {
          throw new Error('Failed to get environment paths');
        }

        envPath = envPathResult.path;
        archivePath = archivePathResult.path;
      }

      // Update progress
      this.updateTask(task.id, { progress: 'Starting training...' });

      // Start training via HTTP API (returns immediately with job ID)
      const trainingResponse = await fetch('http://localhost:8000/training/run', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          config_path: tempConfigPath,
          env_path: envPath,
          archive_path: archivePath,
          job_id: processId
        })
      });

      const startResult = await trainingResponse.json();

      if (startResult.status === 'started') {
        const jobId = startResult.job_id;
        this.updateTask(task.id, { progress: 'Training running...', processId: jobId });

        // Poll for completion
        const finalResult = await this.pollTrainingStatus(jobId, task.id);

        if (finalResult.status === 'completed') {
          return {
            success: true,
            model_save_path: configData.model_save_path,
            job_folder: configData.job_folder,
            config_path: configData.config_output_path,
            classes_trained: configData.class_list.length,
            message: 'Training completed successfully',
            job_id: jobId
          };
        } else if (finalResult.status === 'cancelled') {
          return {
            success: false,
            cancelled: true,
            message: 'Training was cancelled by user',
            job_id: jobId
          };
        } else {
          // Show detailed error information
          let errorMessage = finalResult.error || 'Training failed';

          if (finalResult.stderr) {
            errorMessage += '\n\nError output:\n' + finalResult.stderr;
          }

          if (finalResult.stdout) {
            errorMessage += '\n\nStandard output:\n' + finalResult.stdout;
          }

          throw new Error(errorMessage);
        }
      } else {
        throw new Error(startResult.error || 'Failed to start training');
      }
    } catch (error) {
      throw new Error(`Training failed: ${error.message}`);
    }
  }

  async pollTrainingStatus(jobId, taskId, pollInterval = 2000) {
    /**
     * Poll training status until completion or failure
     * @param {string} jobId - Backend job ID 
     * @param {string} taskId - Frontend task ID
     * @param {number} pollInterval - Polling interval in milliseconds
     * @returns {Promise} Final result when training completes
     */
    return new Promise((resolve, reject) => {
      const poll = async () => {
        try {
          const response = await fetch(`http://localhost:8000/training/status/${jobId}`);

          if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${response.statusText}`);
          }

          const result = await response.json();

          // Update task progress based on status
          if (result.status === 'running') {
            // Store system PID if available
            const updates = { progress: 'Training running...' };
            if (result.system_pid) {
              updates.systemPid = result.system_pid;
            }
            this.updateTask(taskId, updates);
            // Continue polling
            setTimeout(poll, pollInterval);
          } else if (result.status === 'completed') {
            this.updateTask(taskId, { progress: 'Training completed' });
            resolve(result);
          } else if (result.status === 'failed') {
            this.updateTask(taskId, { progress: 'Training failed' });
            resolve(result); // Resolve with failed status, let caller handle error
          } else if (result.status === 'cancelled') {
            this.updateTask(taskId, { progress: 'Cancelled by user' });
            resolve(result); // Resolve with cancelled status - stop polling
          } else {
            // Unknown status, treat as error
            reject(new Error(`Unknown training status: ${result.status}`));
          }
        } catch (error) {
          reject(new Error(`Failed to check training status: ${error.message}`));
        }
      };

      // Start polling
      poll();
    });
  }

  async cancelTask(taskId) {
    const task = this.getTask(taskId);
    if (!task) return false;

    if (task.status === TASK_STATUS.RUNNING) {
      // Cancel the running process via backend API
      if (task.processId) {
        try {
          const endpoint = task.type === TASK_TYPE.TRAINING
            ? `http://localhost:8000/training/cancel/${task.processId}`
            : `http://localhost:8000/inference/cancel/${task.processId}`;

          const response = await fetch(endpoint, {
            method: 'POST',
            headers: {
              'Content-Type': 'application/json',
            }
          });

          if (!response.ok) {
            console.error(`Failed to cancel ${task.type} job via backend:`, response.statusText);
          }
        } catch (error) {
          console.error(`Error cancelling ${task.type} job via backend:`, error);
        }
      }

      // Fallback to Electron API if available (for compatibility)
      if (task.systemPid && window.electronAPI) {
        try {
          window.electronAPI.killPythonProcess(task.systemPid);
        } catch (error) {
          console.error('Error killing process via Electron API:', error);
        }
      }
    }

    // Remove from queue
    this.queue = this.queue.filter(id => id !== taskId);

    // Update task status
    this.updateTask(taskId, {
      status: TASK_STATUS.CANCELLED,
      completed: Date.now(),
      progress: 'Cancelled by user'
    });

    // If this was the current task, process next
    if (this.currentTask === taskId) {
      this.currentTask = null;
      setTimeout(() => this.processQueue(), 100);
    }

    return true;
  }

  // Generate unique job folder name
  async generateUniqueJobFolderName(baseDir, taskName) {
    // Clean the task name for use as folder name
    const cleanedName = taskName.replace(/[^a-zA-Z0-9\-_\s]/g, '').replace(/\s+/g, '_');
    
    // Use Electron API to generate unique folder name
    if (window.electronAPI && window.electronAPI.generateUniqueFolderName) {
      try {
        return await window.electronAPI.generateUniqueFolderName(baseDir, cleanedName);
      } catch (error) {
        console.warn('Failed to generate unique folder name via Electron API, using fallback:', error);
      }
    }
    
    // Fallback: just return the cleaned name (original behavior)
    return cleanedName;
  }

  // Utility Methods
  generateTaskName(config, taskType = TASK_TYPE.INFERENCE) {
    let modelName;
    if (config.model_source === 'bmz') {
      // Use model name from BMZ config
      modelName = config.model || 'Unknown';
    } else{
      modelName = "Local Model"
    }
    const timestamp = new Date().toLocaleString();

    if (taskType === TASK_TYPE.TRAINING) {
      // Generate training task name
      const classCount = Array.isArray(config.class_list) ? config.class_list.length :
        (config.class_list ? config.class_list.split(/[,\n]/).filter(c => c.trim()).length : 0);
      const classDescription = classCount > 0 ? `${classCount} classes` : 'no classes';
      return `Training ${modelName} - ${classDescription} - ${timestamp}`;
    } else {
      // Generate inference task names
      let fileDescription = '';
      if (config.files && config.files.length > 0) {
        fileDescription = `${config.files.length} files`;
      } else if (config.file_globbing_patterns && config.file_globbing_patterns.length > 0) {
        fileDescription = 'pattern-based';
      } else if (config.file_list) {
        fileDescription = 'file list';
      } else {
        fileDescription = 'no files';
      }

      return `Inference ${modelName} - ${fileDescription} - ${timestamp}`;
    }
  }

  // Persistence
  saveTasks() {
    try {
      const tasksData = Array.from(this.tasks.entries());
      localStorage.setItem('inference_tasks', JSON.stringify(tasksData));
    } catch (error) {
      console.error('Failed to save tasks:', error);
    }
  }

  loadTasks() {
    try {
      const tasksData = localStorage.getItem('inference_tasks');
      if (tasksData) {
        const entries = JSON.parse(tasksData);
        this.tasks = new Map(entries);

        // Rebuild queue from queued tasks
        this.rebuildQueue();
      }
    } catch (error) {
      console.error('Failed to load tasks:', error);
      this.tasks = new Map();
    }
  }

  rebuildQueue() {
    // On startup, rebuild queue from tasks that were queued or running
    const queuedTasks = this.getAllTasks()
      .filter(task => task.status === TASK_STATUS.QUEUED || task.status === TASK_STATUS.RUNNING)
      .sort((a, b) => a.created - b.created);

    this.queue = queuedTasks.map(task => {
      // Reset running tasks to queued on startup
      if (task.status === TASK_STATUS.RUNNING) {
        this.updateTask(task.id, {
          status: TASK_STATUS.QUEUED,
          progress: 'Reconnected - queued for execution',
          started: null  // Clear started timestamp so it gets a fresh start time
        });
      }
      return task.id;
    });

    // Start processing if there are queued tasks
    if (this.queue.length > 0) {
      setTimeout(() => this.processQueue(), 1000);
    }
  }

  // Event System
  addListener(callback) {
    this.listeners.add(callback);
    return () => this.listeners.delete(callback);
  }

  notifyListeners(event, data) {
    this.listeners.forEach(callback => {
      try {
        callback(event, data);
      } catch (error) {
        console.error('Task listener error:', error);
      }
    });
  }

  // Queue Info
  getQueueInfo() {
    // Only show currentTask if it's actually still running
    const currentTask = this.currentTask ? this.getTask(this.currentTask) : null;
    const isCurrentTaskRunning = currentTask && currentTask.status === TASK_STATUS.RUNNING;

    return {
      currentTask: isCurrentTaskRunning ? currentTask : null,
      queueLength: this.queue.length,
      nextTasks: this.queue.slice(0, 3).map(id => this.getTask(id))
    };
  }
}

// Singleton instance
export const taskManager = new TaskManager();
export default taskManager;

File: tabs/task_manager.py
"""Task Management for NiceGUI App - Direct Subprocess Execution"""

import asyncio
import json
import uuid
import subprocess
import threading
from pathlib import Path
from typing import Dict, List, Optional
from datetime import datetime
from nicegui import ui


class TaskStatus:
    """Task status constants"""

    UNSTARTED = "unstarted"
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class TaskManager:
    """Manages inference/training/extraction tasks via direct subprocess execution"""

    def __init__(self):
        self.tasks: Dict[str, dict] = {}
        self.current_task = None
        self.queue = []
        self.processes: Dict[str, subprocess.Popen] = {}

    def create_task(self, task_type: str, config: dict, name: str = None) -> dict:
        """Create a new task"""
        task_id = str(uuid.uuid4())

        task = {
            "id": task_id,
            "type": task_type,
            "name": name
            or f'{task_type}_task_{datetime.now().strftime("%Y%m%d_%H%M%S")}',
            "status": TaskStatus.UNSTARTED,
            "config": config,
            "created": datetime.now().isoformat(),
            "started": None,
            "completed": None,
            "progress": "",
            "result": None,
            "job_id": None,
            "system_pid": None,
        }

        self.tasks[task_id] = task
        return task

    def get_task(self, task_id: str) -> Optional[dict]:
        """Get task by ID"""
        return self.tasks.get(task_id)

    def get_all_tasks(self) -> List[dict]:
        """Get all tasks sorted by creation date"""
        return sorted(self.tasks.values(), key=lambda t: t["created"], reverse=True)

    def update_task(self, task_id: str, updates: dict):
        """Update task fields"""
        if task_id in self.tasks:
            self.tasks[task_id].update(updates)

    async def queue_task(self, task_id: str):
        """Add task to queue and start processing"""
        task = self.get_task(task_id)
        if not task or task["status"] == TaskStatus.RUNNING:
            return False

        self.update_task(
            task_id, {"status": TaskStatus.QUEUED, "progress": "Queued for execution"}
        )

        if task_id not in self.queue:
            self.queue.append(task_id)

        # Start processing if no current task
        if not self.current_task:
            await self.process_queue()

        return True

    async def process_queue(self):
        """Process tasks in queue"""
        if self.current_task or not self.queue:
            return

        task_id = self.queue.pop(0)
        task = self.get_task(task_id)

        if not task or task["status"] == TaskStatus.CANCELLED:
            # Task was deleted or cancelled, process next
            await self.process_queue()
            return

        self.current_task = task_id
        await self.execute_task(task_id)

    async def execute_task(self, task_id: str):
        """Execute a task via subprocess"""
        task = self.get_task(task_id)
        if not task:
            return

        try:
            self.update_task(
                task_id,
                {
                    "status": TaskStatus.RUNNING,
                    "started": datetime.now().isoformat(),
                    "progress": f'Starting {task["type"]}...',
                },
            )

            # Execute based on task type
            if task["type"] == "inference":
                await self.run_inference(task)
            elif task["type"] == "training":
                await self.run_training(task)
            elif task["type"] == "extraction":
                await self.run_extraction(task)

        except Exception as e:
            print(f"Task execution error: {e}")
            import traceback

            traceback.print_exc()
            self.update_task(
                task_id,
                {
                    "status": TaskStatus.FAILED,
                    "completed": datetime.now().isoformat(),
                    "progress": "Failed",
                    "result": {"error": str(e)},
                },
            )
        finally:
            self.current_task = None
            # Process next task
            await self.process_queue()

    async def run_inference(self, task: dict):
        """Run inference task via direct subprocess"""
        config = task["config"]

        # Create job folder name
        job_folder_name = f"{task['name']}_{task['id'][:8]}"
        output_dir = config.get("output_dir", "/tmp")
        job_folder = Path(output_dir) / job_folder_name
        job_folder.mkdir(parents=True, exist_ok=True)

        # Prepare config for inference script
        # Handle sparse_save_threshold: "none" when disabled, otherwise the numeric value
        if config.get("sparse_outputs_enabled", False):
            sparse_threshold = config.get("sparse_save_threshold", -3.0)
        else:
            sparse_threshold = "none"
        
        inference_config = {
            "model_source": config.get("model_source", "bmz"),
            "mode": "classification",  # 'classification', 'embed_to_hoplite', and 'classify_from_hoplite'
            "model": config.get("model", "BirdSetEfficientNetB1"),
            "files": config.get("files", []),
            "file_globbing_patterns": config.get("file_globbing_patterns", []),
            "file_list": config.get("file_list", ""),
            "job_folder": str(job_folder),
            "sparse_save_threshold": sparse_threshold,
            "split_by_subfolder": config.get("split_by_subfolder", False),
            "subset_size": (
                config.get("subset_size")
                if config.get("testing_mode_enabled")
                else None
            ),
            "inference_settings": {
                "clip_overlap": config.get("overlap", 0.0),
                "batch_size": config.get("batch_size", 1),
                "num_workers": config.get("worker_count", 1),
            },
        }

        # Save config file
        config_path = job_folder / f'{task["name"]}_config.json'
        config_path.write_text(json.dumps(inference_config, indent=2))

        # Log file path
        log_file_path = job_folder / "inference_log.txt"

        # Determine Python environment
        if config.get("use_custom_python_env") and config.get("custom_python_env_path"):
            python_path = Path(config["custom_python_env_path"])
            if python_path.is_dir():
                # If it's a directory, assume it's a conda/venv environment
                python_exe = python_path / "bin" / "python"
            else:
                python_exe = python_path
        else:
            # Default location of Python executable installed by Dipper
            # (change later to match install location per system)
            # env should be downloaded from url and unzipped if not present
            python_exe = Path(
                "/Users/SML161/Library/Application Support/Electron/envs/dipper_pytorch_env"
            )

        # Path to inference script
        script_path = (
            Path(__file__).parent.parent / "backend" / "scripts" / "inference.py"
        )

        try:
            # Build command
            cmd = [str(python_exe), str(script_path), "--config", str(config_path)]

            print(f'Starting inference task: {" ".join(cmd)}')

            # Start subprocess
            log_file = open(log_file_path, "w")
            process = subprocess.Popen(
                cmd,
                stdout=log_file,
                stderr=subprocess.STDOUT,
                text=True,
                cwd=str(script_path.parent),
            )

            # Store process
            self.processes[task["id"]] = process

            self.update_task(
                task["id"],
                {
                    "system_pid": process.pid,
                    "progress": f"Running inference (PID: {process.pid})...",
                    "log_file": str(log_file_path),
                    "job_folder": str(job_folder),
                },
            )

            # Monitor process in background thread
            def monitor_process():
                try:
                    return_code = process.wait()
                    log_file.close()

                    if return_code == 0:
                        self.update_task(
                            task["id"],
                            {
                                "status": TaskStatus.COMPLETED,
                                "completed": datetime.now().isoformat(),
                                "progress": "Completed successfully",
                                "result": {"exit_code": return_code},
                            },
                        )
                        ui.notify(
                            f'Task "{task["name"]}" completed successfully',
                            type="positive",
                        )
                    else:
                        # Read last lines of log file for error
                        try:
                            with open(log_file_path, "r") as f:
                                log_lines = f.readlines()
                                error_msg = (
                                    "".join(log_lines[-10:])
                                    if log_lines
                                    else f"Exit code: {return_code}"
                                )
                        except:
                            error_msg = f"Exit code: {return_code}"

                        self.update_task(
                            task["id"],
                            {
                                "status": TaskStatus.FAILED,
                                "completed": datetime.now().isoformat(),
                                "progress": "Failed",
                                "result": {
                                    "error": error_msg,
                                    "exit_code": return_code,
                                },
                            },
                        )
                        ui.notify(f'Task "{task["name"]}" failed', type="negative")

                    # Remove from processes dict
                    if task["id"] in self.processes:
                        del self.processes[task["id"]]

                except Exception as e:
                    print(f"Error monitoring process: {e}")
                    self.update_task(
                        task["id"],
                        {
                            "status": TaskStatus.FAILED,
                            "completed": datetime.now().isoformat(),
                            "progress": "Failed",
                            "result": {"error": str(e)},
                        },
                    )

            # Start monitoring thread
            monitor_thread = threading.Thread(target=monitor_process, daemon=True)
            monitor_thread.start()

        except Exception as e:
            print(f"Failed to start inference: {e}")
            import traceback

            traceback.print_exc()
            self.update_task(
                task["id"],
                {
                    "status": TaskStatus.FAILED,
                    "completed": datetime.now().isoformat(),
                    "progress": "Failed to start",
                    "result": {"error": str(e)},
                },
            )
            raise

    async def run_training(self, task: dict):
        """Run training task"""
        # TODO: Implement training similar to inference
        ui.notify("Training not yet implemented", type="warning")
        self.update_task(
            task["id"],
            {
                "status": TaskStatus.FAILED,
                "completed": datetime.now().isoformat(),
                "progress": "Not implemented",
                "result": {"error": "Training not yet implemented"},
            },
        )

    async def run_extraction(self, task: dict):
        """Run extraction task"""
        # TODO: Implement extraction similar to inference
        ui.notify("Extraction not yet implemented", type="warning")
        self.update_task(
            task["id"],
            {
                "status": TaskStatus.FAILED,
                "completed": datetime.now().isoformat(),
                "progress": "Not implemented",
                "result": {"error": "Extraction not yet implemented"},
            },
        )

    async def cancel_task(self, task_id: str):
        """Cancel a running task by terminating the subprocess"""
        task = self.get_task(task_id)
        if not task:
            return False

        # Get the process
        process = self.processes.get(task_id)
        if process and process.poll() is None:  # Process is still running
            try:
                process.terminate()
                # Give it a moment to terminate gracefully
                await asyncio.sleep(0.5)
                if process.poll() is None:
                    # Force kill if still running
                    process.kill()

                self.update_task(
                    task_id,
                    {
                        "status": TaskStatus.CANCELLED,
                        "completed": datetime.now().isoformat(),
                        "progress": "Cancelled by user",
                    },
                )

                if task_id in self.processes:
                    del self.processes[task_id]

                ui.notify(f'Task "{task["name"]}" cancelled', type="warning")
                return True
            except Exception as e:
                print(f"Error cancelling task: {e}")
                return False

        return False

    def delete_task(self, task_id: str):
        """Delete a task"""
        task = self.get_task(task_id)
        if task and task["status"] != TaskStatus.RUNNING:
            del self.tasks[task_id]
            if task_id in self.queue:
                self.queue.remove(task_id)
            return True
        return False


# Global task manager instance
task_manager = TaskManager()
